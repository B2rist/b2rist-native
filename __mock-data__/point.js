export const points = [
  {
    user: 'test@gmail.com',
    contentType: 'sound-collage',
    name: 'Elephants',
    description:
      'Elephants are the heaviest land mammals in the world. The male African elephant can reach a weight of up to 6 tonnes. The Asian elephant, smaller than it’s African relative, has been domesticated for thousands of years, often being used as pack-animals. Katka and Wiki live in Oliwa Zoo and are rather elderly representatives of both species. They live together in the pavilion, which was prepared for the elephants in the early 50s.  The African elephant Katka was born around 1970 and came to us from France. She presented her excellent memory by recognizing her Czech keeper after not seeing him for 16 years!  The Asian elephant Wiki was born in 1964 and came to Oliwa zoo under very tragic circumstances after animal rights protection organisation retrieved her from a closed circus. She found her new home here with us in Gdansk. After the traumatic experience of working in a circus, Wiki needed two years in order to trust people again',
    radius: 20,
    isRoute: false,
    id: 'm3Ss7dha9TRCT8na9irS',
    location: {
      _latitude: 54.41221902644764,
      _longitude: 18.533400003253522,
    },
    cover: '1QxEYG2Z_yUrltO2S02aZC3q39WtlO1bj',
    geohash: 'u3tjt9rw5s',
    language: 'en',
    audioContent: '1SPT9SwLvr89EFdxydfZ7yQ4d_jEE9yvd',
    images: [
      {
        time: 0,
        image: '1QxEYG2Z_yUrltO2S02aZC3q39WtlO1bj',
      },
    ],
  },
  {
    user: 'test@gmail.com',
    contentType: 'sound-collage',
    name: 'Lions',
    description:
      'Lives in southwest Africa and is the biggest predator of the continent. It can be found in various environments: forested, semi-desert and even mountainous areas at the altitude of 4000m. Unlike other wild cats, lions live in pride consisting of up to 30 individuals. Dominant male is responsible for protection of offspring and his territory while females prey for food and care for their cubs. Most of their time lions spend resting and sleeping. Although lions are able to swim they avoid water.',
    radius: 20,
    isRoute: false,
    id: '0wcA2uvk3yVHBrqpRKwB',
    location: {
      _latitude: 54.410370053735996,
      _longitude: 18.539548069238663,
    },
    cover: '1N3cjoN3DBovGiySj6IwHCclBdyFH6Lks',
    geohash: 'u3tjtch5cq',
    language: 'en',
    audioContent: '1SPT9SwLvr89EFdxydfZ7yQ4d_jEE9yvd',
    images: [
      {
        time: 0,
        image: '1N3cjoN3DBovGiySj6IwHCclBdyFH6Lks',
      },
    ],
  },
  {
    user: 'test@gmail.com',
    contentType: 'sound-collage',
    name: 'Green gate',
    description:
      "With the Golden Gate and the Highland Gate, the Green Gate spans the Long Market and Long Street, together comprising the Royal Route. The Green Gate was clearly inspired by the Antwerp City Hall.[1] It was built 1568-71 as the formal residence of Poland's monarchs.[2] It is a masterpiece by Regnier (or Reiner van Amsterdam), an Amsterdam architect,[3] and reflects Flemish architectural influence in Gdańsk. Hans Kramer from Dresden was responsible for the construction plans.  On 11–20 February 1646 the future Queen of Poland, Marie Louise Gonzaga, was entertained here. In the late 18th century the Nature Society was housed here, but soon moved to the Naturalists' House (Research Society House).  Today the Green Gate houses the National Museum in Gdańsk. Exhibitions, meetings, conferences and shows are held here. The Gdańsk office of former Polish President Lech Wałęsa is located in one of the rooms.",
    radius: 20,
    isRoute: false,
    id: 'L5Bq5UzXl1EkAKj9E4xf',
    location: {
      _latitude: 54.34817976567255,
      _longitude: 18.65583504157628,
    },
    cover: '1Dr1ml2-GZcA5zyf68VmiFudFAoyCu9eb',
    geohash: 'u3tm0tbc07',
    language: 'en',
    audioContent: '1SPT9SwLvr89EFdxydfZ7yQ4d_jEE9yvd',
    images: [
      {
        time: 0,
        image: '1Dr1ml2-GZcA5zyf68VmiFudFAoyCu9eb',
      },
      {
        time: 10,
        image: '1JXNtJrUOfYppjEpzIBgQ7EWTs31Bs-Dy',
      },
    ],
  },
  {
    user: 'test@gmail.com',
    contentType: 'sound-collage',
    name: 'Stocznia Gdańska',
    description:
      'Brama nr 2 Stoczni Gdańskiej – jedna z bram prowadzących do Stoczni Gdańskiej. Ze względu na bliskość budynków dyrekcji stoczni oraz dogodne połączenie ze Śródmieściem Gdańska i dworcem kolejowym Gdańsk Główny powszechnie uznawana za główne wejście na teren zakładu.  W 1999 brama została wpisana do rejestru zabytków województwa pomorskiego pod numerem A-1206, a w 2014 została wyróżniona Znakiem Dziedzictwa Europejskiego w ramach historycznego zespołu obiektów Stoczni Gdańskiej',
    radius: 20,
    isRoute: false,
    id: 'C8U9O6WC9YQIsKVgpFzm',
    location: {
      _latitude: 54.360587751617885,
      _longitude: 18.649858281381743,
    },
    cover: '15MA3J9bHClcGGhoZBScMVVi3bxoS5_U3',
    geohash: 'u3tm22h3mh',
    language: 'en',
    audioContent: '1SPT9SwLvr89EFdxydfZ7yQ4d_jEE9yvd',
    images: [
      {
        time: 0,
        image: '15MA3J9bHClcGGhoZBScMVVi3bxoS5_U3',
      },
      {
        time: 10,
        image: '1qwvpjIsdsC-Nw1zP2o9lPi_6ZiGgBHYI',
      },
    ],
  },
];
