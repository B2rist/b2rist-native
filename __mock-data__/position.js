export const position = {
  speed: 0,
  heading: 0,
  altitude: 0,
  accuracy: 5.258999824523926,
  longitude: 18.65263,
  altitudeAccuracy: 0.5,
  latitude: 54.3486967,
};
