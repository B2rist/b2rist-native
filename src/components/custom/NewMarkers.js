import React, {useCallback} from 'react';
import {Marker} from 'react-native-maps';
import {MARKER} from 'images';
import PropTypes from 'prop-types';

const NewMarkers = ({points, setPoint}) => {
  const onSetPoint = useCallback(point => () => setPoint(point), [setPoint]);
  return points.map(point => (
    <Marker
      key={point.id}
      title={point.name}
      onPress={onSetPoint(point)}
      image={MARKER}
      coordinate={{
        latitude: point.location._latitude,
        longitude: point.location._longitude,
      }}
    />
  ));
};

const Memo = React.memo(NewMarkers);

export default Memo;

Memo.propTypes = {
  points: PropTypes.array.isRequired,
  setPoint: PropTypes.func,
};

Memo.defaultProps = {
  setPoint: () => {},
};
