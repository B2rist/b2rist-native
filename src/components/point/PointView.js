import React, {useCallback} from 'react';
import {Avatar, Button, Card, Paragraph} from 'react-native-paper';
import PanelHandle from 'components/point/PanelHandle';
import PropTypes from 'prop-types';

const MapMarker = props => <Avatar.Icon {...props} icon="map-marker" />;

const PointView = ({
  point,
  distance,
  isNearby,
  setPlayingPoint,
  clearPoint,
}) => {
  const onSetPlayingPoint = useCallback(() => {
    setPlayingPoint(point);
  }, [point, setPlayingPoint]);

  return (
    <Card>
      <Card.Content>
        <PanelHandle />
      </Card.Content>
      <Card.Title
        title={point.name}
        subtitle={`${distance} m`}
        left={MapMarker}
      />
      <Card.Actions>
        {!isNearby && (
          <Button icon="google-nearby" mode="outlined" onPress={clearPoint}>
            Show nearby
          </Button>
        )}
        <Button
          icon="play-box-outline"
          mode="contained"
          onPress={onSetPlayingPoint}>
          Play
        </Button>
      </Card.Actions>
      {point.cover && <Card.Cover source={{uri: point.cover}} />}
      <Card.Content>
        <Paragraph>{point.description}</Paragraph>
      </Card.Content>
    </Card>
  );
};

export default PointView;

PointView.propTypes = {
  point: PropTypes.shape({
    name: PropTypes.string.isRequired,
    cover: PropTypes.string,
    description: PropTypes.string,
  }).isRequired,
  distance: PropTypes.number.isRequired,
  isNearby: PropTypes.bool.isRequired,
  setPlayingPoint: PropTypes.func.isRequired,
  clearPoint: PropTypes.func.isRequired,
};
