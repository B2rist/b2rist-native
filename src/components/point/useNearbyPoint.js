import {useEffect, useState} from 'react';
import {getDistance} from 'geolib';

// TODO: Move to the firebase #8
const pointArea = 20;

const useNearbyPoint = (points, position, gpsEnabled) => {
  const [point, setPoint] = useState(null);
  const [index, setIndex] = useState(null);
  const [reached, setReached] = useState(false);
  const [distance, setDistance] = useState(null);
  useEffect(() => {
    if (gpsEnabled && points) {
      const closeIn = points
        .map((item, ind) => {
          const location = {
            latitude: item.location._latitude,
            longitude: item.location._longitude,
          };
          return {
            point: item,
            index: ind,
            distance: getDistance(position, location),
          };
        })
        .sort((a, b) => b.distance - a.distance)
        .pop();
      point?.id !== closeIn?.point.id &&
        setPoint(closeIn.point) &&
        setIndex(closeIn.index);
      closeIn?.distance &&
        closeIn.distance !== distance &&
        setDistance(closeIn?.distance);
      point && distance < pointArea && setReached(true);
      point && distance > pointArea && setReached(false);
    }
  }, [distance, gpsEnabled, point, points, position]);
  return [point, distance, reached, index];
};

export default useNearbyPoint;
