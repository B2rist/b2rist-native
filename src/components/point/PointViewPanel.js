import React, {useCallback, useMemo, useRef, useState} from 'react';
import {
  Animated,
  Platform,
  ScrollView,
  StyleSheet,
  useWindowDimensions,
  View,
} from 'react-native';
import SlidingUpPanel from 'rn-sliding-up-panel';
import PointView from 'components/point/PointView';
import useDarkMode from 'components/custom/useDarkMode';
import {Colors} from 'react-native/Libraries/NewAppScreen';
import PropTypes from 'prop-types';

const ios = Platform.OS === 'ios';

const PointViewPanel = ({
  point,
  distance,
  isNearby,
  mapDimension,
  setPlayingPoint,
  clearPoint,
}) => {
  const darkMode = useDarkMode();
  const styles = useMemo(() => makeStyles(darkMode), [darkMode]);
  const deviceHeight = useWindowDimensions().height;

  const draggableRange = useMemo(
    () => ({
      top: mapDimension.height + 1,
      bottom: deviceHeight / 7.8,
    }),
    [deviceHeight, mapDimension.height],
  );

  const [scrollEnabled, setScrollEnabled] = useState(false);
  const [allowDragging, setAllowDragging] = useState(true);
  const [atTop, setAtTop] = useState(true);

  const snappingPoints = [draggableRange.top, draggableRange.bottom];

  const panelRef = useRef();
  const [panelPositionVal] = useState(
    new Animated.Value(draggableRange.bottom),
  );

  const onMomentumDragEnd = useCallback(
    (value: number) => {
      if (value === draggableRange.top && !scrollEnabled) {
        setScrollEnabled(true);
        setAtTop(true);
      }
    },
    [draggableRange, scrollEnabled],
  );

  const onMomentumScrollEnd = useCallback(event => {
    const {nativeEvent} = event;
    if (nativeEvent.contentOffset.y === 0) {
      setAtTop(true);
      if (ios) {
        setAllowDragging(true);
      }
    }
  }, []);

  const PANEL_VELOCITY = ios ? 1 : 2.3;
  const hideFullScreenPanelOptions = useMemo(
    () => ({
      velocity: PANEL_VELOCITY,
      toValue: draggableRange.bottom,
    }),
    [PANEL_VELOCITY, draggableRange.bottom],
  );

  const onDragStart = useCallback(
    (_, gestureState) => {
      if (atTop && scrollEnabled) {
        if (gestureState.vy > 0) {
          setScrollEnabled(false);
          if (ios) {
            setAllowDragging(true);
          }
          if (panelRef && panelRef.current) {
            panelRef.current.show(hideFullScreenPanelOptions);
          }
        } else {
          setAtTop(false);
          if (ios) {
            setAllowDragging(false);
          }
        }
      }
    },
    [atTop, scrollEnabled, hideFullScreenPanelOptions],
  );
  return (
    <SlidingUpPanel
      ref={panelRef}
      animatedValue={panelPositionVal}
      draggableRange={draggableRange}
      snappingPoints={snappingPoints}
      backdropOpacity={0}
      showBackdrop={false}
      height={mapDimension.height}
      allowDragging={allowDragging}
      onMomentumDragEnd={onMomentumDragEnd}
      onDragStart={onDragStart}>
      <View style={styles.panelContent}>
        <ScrollView
          scrollEnabled={scrollEnabled}
          showsVerticalScrollIndicator={false}
          bounces={false}
          onMomentumScrollEnd={onMomentumScrollEnd}>
          <PointView
            clearPoint={clearPoint}
            isNearby={isNearby}
            point={point}
            distance={distance}
            setPlayingPoint={setPlayingPoint}
          />
        </ScrollView>
      </View>
    </SlidingUpPanel>
  );
};

const makeStyles = isDarkMode =>
  StyleSheet.create({
    panelContent: {
      flex: 1,
      width: '100%',
      height: '100%',
      backgroundColor: isDarkMode ? Colors.darker : Colors.lighter,
    },
  });

export default PointViewPanel;

PointViewPanel.propTypes = {
  point: PropTypes.object.isRequired,
  distance: PropTypes.number.isRequired,
  isNearby: PropTypes.bool.isRequired,
  mapDimension: PropTypes.shape({
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
  }).isRequired,
  setPlayingPoint: PropTypes.func.isRequired,
  clearPoint: PropTypes.func.isRequired,
};
