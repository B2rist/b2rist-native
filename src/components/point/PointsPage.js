import React, {
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';
import {useSelector} from 'react-redux';
import {selectPointPage} from 'services/redux/pointSlice';
import {Platform, Slider, StyleSheet, View} from 'react-native';
import GoogleMapView from 'components/googlemaps/GoogleMapView';
import {GeoLocationContext} from 'GeoLocationProvider';
import {
  Banner,
  Button,
  Card,
  IconButton,
  Avatar,
  Modal,
  Portal,
} from 'react-native-paper';
import MediaResolver from 'components/player/MediaResolver';
import {useNavigate, createSearchParams} from 'react-router-native';
import useNearbyPoint from 'components/point/useNearbyPoint';
import NewMarkers from 'components/custom/NewMarkers';

const MapMarker = props => <Avatar.Icon {...props} icon="map-marker" />;

const PlayButton = (props, onPress) => (
  <IconButton {...props} icon="play-box-outline" onPress={onPress} />
);

const PointsPage = () => {
  const navigate = useNavigate();
  const points = useSelector(state => selectPointPage(state));
  const {position, enabled: gpsEnabled} = useContext(GeoLocationContext);
  const [playingPoint, setPlayingPoint] = useState(null);
  const [played, setPlayed] = useState([]);
  const [filterVisible, setFilterVisible] = useState(false);
  const [nearbyPoint, distance, reached] = useNearbyPoint(
    points.list,
    position,
    gpsEnabled,
  );

  useEffect(() => {
    nearbyPoint &&
      reached &&
      playingPoint?.id !== nearbyPoint.id &&
      !played.includes(nearbyPoint.id) &&
      setPlayingPoint(nearbyPoint);
  }, [nearbyPoint, played, playingPoint, reached]);

  const onClose = useCallback(() => {
    setPlayed([...played, playingPoint.id]);
    setPlayingPoint(null);
  }, [played, playingPoint]);

  const onValueChanged = useCallback(
    value =>
      navigate({
        pathname: '/points',
        search: `?${createSearchParams({radius: value})}`,
      }),
    [navigate],
  );

  const onSetPlayingPoint = useCallback(point => {
    setPlayingPoint(point);
  }, []);

  const Markers = useMemo(
    () => (
      <NewMarkers points={points.list} setPlayingPoint={onSetPlayingPoint} />
    ),
    [onSetPlayingPoint, points],
  );

  return (
    <>
      <Portal>
        <Modal
          visible={!!playingPoint}
          onDismiss={onClose}
          contentContainerStyle={styles.modal}>
          {playingPoint && (
            <MediaResolver point={playingPoint} close={onClose} />
          )}
        </Modal>
      </Portal>
      <Banner
        visible={filterVisible}
        actions={[
          {
            label: 'Close',
            onPress: () => setFilterVisible(false),
          },
        ]}>
        <Slider
          onSlidingComplete={onValueChanged}
          value={+points.page.radius}
          minimumValue={0}
          maximumValue={25}
          maximumTrackTintColor="gray"
          minimumTrackTintColor={'white'}
          thumbTintColor={'white'}
          style={styles.slider}
        />
      </Banner>
      {!filterVisible && (
        <View style={styles.moreButton}>
          <Button onPress={() => setFilterVisible(true)}>... More</Button>
        </View>
      )}
      <View style={styles.container}>
        <GoogleMapView children={Markers} />
      </View>
      {nearbyPoint && (
        <Card>
          <Card.Title
            title={nearbyPoint.name}
            subtitle={`${distance} m`}
            left={MapMarker}
            right={props =>
              PlayButton(props, () => setPlayingPoint(nearbyPoint))
            }
          />
        </Card>
      )}
    </>
  );
};

const styles = StyleSheet.create({
  modal: {
    marginTop: -20,
    flex: 1,
  },
  moreButton: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
  },
  container: {
    marginTop: 10,
    height: 400,
  },
  slider: {
    width: 300,
    height: 30,
    borderRadius: 50,
    flex: 1,
    alignSelf: 'center',
    marginHorizontal: Platform.select({ios: 5}),
  },
});

export default PointsPage;
