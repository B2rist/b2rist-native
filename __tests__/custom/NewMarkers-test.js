import 'react-native';
import React from 'react';
import {fireEvent, render, screen} from '@testing-library/react-native';
import {points} from '../../__mock-data__/point';
import NewMarkers from 'components/custom/NewMarkers';

const setPointMock = jest.fn();

it('new marker test', () => {
  const {getByTestId} = render(
    <NewMarkers points={points} setPoint={setPointMock} />,
  );
  expect(screen.toJSON()).toMatchSnapshot();
  fireEvent.press(getByTestId(`new-marker-${points[0].id}`));
  expect(setPointMock).toHaveBeenCalledTimes(1);
  expect(setPointMock).toHaveBeenCalledWith(points[0]);
});
