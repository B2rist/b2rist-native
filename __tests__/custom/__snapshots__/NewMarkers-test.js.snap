// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new marker test 1`] = `
[
  <AIRMapMarker
    coordinate={
      {
        "latitude": 54.41221902644764,
        "longitude": 18.533400003253522,
      }
    }
    onPress={[Function]}
    style={
      [
        {
          "backgroundColor": "transparent",
          "position": "absolute",
        },
        undefined,
      ]
    }
    testID="new-marker-m3Ss7dha9TRCT8na9irS"
    title="Elephants"
    tracksViewChanges={false}
  >
    <Image
      source={
        {
          "testUri": "../../../src/images/marker.png",
        }
      }
      style={
        {
          "height": 29,
          "width": 24,
        }
      }
    />
  </AIRMapMarker>,
  <AIRMapMarker
    coordinate={
      {
        "latitude": 54.410370053735996,
        "longitude": 18.539548069238663,
      }
    }
    onPress={[Function]}
    style={
      [
        {
          "backgroundColor": "transparent",
          "position": "absolute",
        },
        undefined,
      ]
    }
    testID="new-marker-0wcA2uvk3yVHBrqpRKwB"
    title="Lions"
    tracksViewChanges={false}
  >
    <Image
      source={
        {
          "testUri": "../../../src/images/marker.png",
        }
      }
      style={
        {
          "height": 29,
          "width": 24,
        }
      }
    />
  </AIRMapMarker>,
  <AIRMapMarker
    coordinate={
      {
        "latitude": 54.34817976567255,
        "longitude": 18.65583504157628,
      }
    }
    onPress={[Function]}
    style={
      [
        {
          "backgroundColor": "transparent",
          "position": "absolute",
        },
        undefined,
      ]
    }
    testID="new-marker-L5Bq5UzXl1EkAKj9E4xf"
    title="Green gate"
    tracksViewChanges={false}
  >
    <Image
      source={
        {
          "testUri": "../../../src/images/marker.png",
        }
      }
      style={
        {
          "height": 29,
          "width": 24,
        }
      }
    />
  </AIRMapMarker>,
  <AIRMapMarker
    coordinate={
      {
        "latitude": 54.360587751617885,
        "longitude": 18.649858281381743,
      }
    }
    onPress={[Function]}
    style={
      [
        {
          "backgroundColor": "transparent",
          "position": "absolute",
        },
        undefined,
      ]
    }
    testID="new-marker-C8U9O6WC9YQIsKVgpFzm"
    title="Stocznia GdaÅ„ska"
    tracksViewChanges={false}
  >
    <Image
      source={
        {
          "testUri": "../../../src/images/marker.png",
        }
      }
      style={
        {
          "height": 29,
          "width": 24,
        }
      }
    />
  </AIRMapMarker>,
]
`;
